File: Final_Project/Code/Project_Interger_Partition/Bai_toan_2/CSS_bt2.cpp
#include <iostream>
#include <vector>
using namespace std;

int count_partitions(int n, int k) {
    vector<vector<int>> dp(n + 1, vector<int>(k + 1, 0));
    dp[0][0] = 1;
    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= min(i, k); j++)
            dp[i][j] = dp[i - j][j] + (j > 1 ? dp[i - 1][j - 1] : 0);
    return dp[n][k];
}

int count_max_partitions(int n, int k) {
    return n >= k ? count_partitions(n - k, k - 1) : 0;
}

int main() {
    int n, k;
    cout << "Nhap n: "; cin >> n;
    cout << "Nhap k: "; cin >> k;
    int pk_n = count_partitions(n, k);
    int pmax_n_k = count_max_partitions(n, k);
    cout << "p_" << k << "(" << n << ") = " << pk_n << endl;
    cout << "p_max(" << n << "," << k << ") = " << pmax_n_k << endl;
    cout << "So sanh: p_max(" << n << "," << k << ") <= p_" << k << "(" << n << ") is " << (pmax_n_k <= pk_n) << endl;
    return 0;
}

File: Final_Project/Code/Project_Interger_Partition/Bai_toan_2/CSS_bt2.py
def count_partitions(n, k):
    dp = [[0] * (k + 1) for _ in range(n + 1)]
    dp[0][0] = 1
    for i in range(1, n + 1):
        for j in range(1, min(i + 1, k + 1)):
            dp[i][j] = dp[i - j][j] + (dp[i - 1][j - 1] if j > 1 else 0)
    return dp[n][k]

def count_max_partitions(n, k):
    return count_partitions(n - k, k - 1) if n >= k else 0

def main():
    n = int(input("Nhap n: "))
    k = int(input("Nhap k: "))
    pk_n = count_partitions(n, k)
    pmax_n_k = count_max_partitions(n, k)
    print(f"p_{k}({n}) = {pk_n}")
    print(f"p_max({n},{k}) = {pmax_n_k}")
    print(f"So sanh: p_max({n},{k}) <= p_{k}({n}) is {pmax_n_k <= pk_n}")

if __name__ == "__main__":
    main()